File: solutions/java/code-carl/ch08_greedy/T8_LC0134_CanCompleteCircuit.java
/**
 * Practice template for:
 * - LeetCode 134. Gas Station
 *   https://leetcode.cn/problems/gas-station/
 */
/**
 * @author Cheng Yihao
 * @version 1.0
 * @date 2025/4/4 22:17
 * @comment
 */
public class T8_LC0134_CanCompleteCircuit {
    public int canCompleteCircuit(int[] gas, int[] cost) {
        int currGas = 0;
        int totalGas = 0;
        int begin = 0;

        for (int i = 0; i < gas.length; i++) {
            totalGas += gas[i] - cost[i];
            currGas += gas[i] - cost[i];
            if (currGas < 0) {
                begin = i + 1;
                currGas = 0;
            }
        }
        return totalGas < 0 ? -1 : begin;
    }
}

File: solutions/code-carl/ch08_greedy/t8_lc0134_can_complete_circuit.py
"""LeetCode 134. Gas Station
https://leetcode.cn/problems/gas-station/
"""

from typing import List


class Solution:
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        start_idx = 0
        total_gas = 0
        curr_gas = 0

        for i, (g, c) in enumerate(zip(gas, cost)):
            total_gas += g - c
            curr_gas += g - c
            if curr_gas < 0:
                start_idx = i + 1
                curr_gas = 0

        return start_idx if total_gas >= 0 else -1


if __name__ == "__main__":
    solution = Solution()
    assert solution.canCompleteCircuit([1, 2, 3, 4, 5], [3, 4, 5, 1, 2]) == 3
    assert solution.canCompleteCircuit([2, 3, 4], [3, 4, 3]) == -1
    print("All tests passed.")
